linkerd/io/src/scoped.rs

```diff
@@ -0,0 +1,133 @@
+use crate as io;
+use pin_project::pin_project;
+use std::{pin::Pin, task::Context};
+
+/// An I/O stream where errors are annotated a scope.
+#[pin_project]
+#[derive(Debug)]
+pub struct ScopedIo<I> {
+    scope: Scope,
+
+    #[pin]
+    io: I,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum Scope {
+    Client,
+    Server,
+}
+
+// === impl Scope ===
+
+impl Scope {
+    fn err(&self, err: io::Error) -> io::Error {
+        io::Error::new(err.kind(), format!("{}: {}", self, err.to_string()))
```

Comment:
nit: the `to_string` _shouldn't_ be necessary --- that's equivalent to `format!("{}", error)`, and we're already making a string, so this is just
```suggestion
        io::Error::new(err.kind(), format!("{}: {}", self, err))
```
but with extra steps